spot_rl_experiments/spot_rl/real_policy.py

```diff
@@ -24,76 +29,160 @@ def to_tensor(v):
         self.test_recurrent_hidden_states = None
         self.not_done_masks = None
         self.config = config
         self.num_actions = action_space.shape[0]
         self.reset_ran = False
+        self.is_hab3_policy = is_hab3_policy
         print("Policy loaded.")
 
     def reset(self):
         self.reset_ran = True
-        self.test_recurrent_hidden_states = torch.zeros(
-            1,  # The number of environments. Just one for real world.
-            self.policy.net.num_recurrent_layers,
-            self.config.RL.PPO.hidden_size,
-            device=self.device,
-        )
+        if not self.is_hab3_policy:
+            self.test_recurrent_hidden_states = torch.zeros(
+                1,  # The number of environments. Just one for real world.
+                self.policy.net.num_recurrent_layers,
+                self.config.RL.PPO.hidden_size,
+                device=self.device,
+            )
+        else:
+            self.test_recurrent_hidden_states = torch.zeros(
+                1,  # The number of environments. Just one for real world.
+                self.config.get("MOBILE_GAZE_NUM_RECURRENT_LAYERS", 4),
+                self.config.get("MOBILE_GAZE_RL_PPO_HIDDEN_SIZE", 512),
+                device=self.device,
+            )
 
         # We start an episode with 'done' being True (0 for 'not_done')
         self.not_done_masks = torch.zeros(1, 1, dtype=torch.bool, device=self.device)
         self.prev_actions = torch.zeros(1, self.num_actions, device=self.device)
 
+    def get_action_distribution(self, mu_maybe_std, std):
+        """The final transformation of the action given inputs"""
+        # We use mu_maybe_std to follow the convension of habitat-baselines
+        mu = torch.tanh(mu_maybe_std.float())
+        std = torch.exp(torch.clamp(std, -5, 2))
+
+        return CustomNormal(mu, std, validate_args=False)
+
     def act(self, observations):
         assert self.reset_ran, "You need to call .reset() on the policy first."
         batch = batch_obs([observations], device=self.device)
         with torch.no_grad():
-            _, actions, _, self.test_recurrent_hidden_states = self.policy.act(
-                batch,
-                self.test_recurrent_hidden_states,
-                self.prev_actions,
-                self.not_done_masks,
-                deterministic=True,
-                actions_only=True,
-            )
+            if not self.is_hab3_policy:
+                _, actions, _, self.test_recurrent_hidden_states = self.policy.act(
+                    batch,
+                    self.test_recurrent_hidden_states,
+                    self.prev_actions,
+                    self.not_done_masks,
+                    deterministic=True,
+                    actions_only=True,
+                )
+            else:
+                # Using torch script to save the model
+                with torch.no_grad():
+                    output_model = self.policy["net"](
```

Comment:
why is this model loaded every policy call? Should this be loaded once and then used here?

Reply:
@akshararai the variable name might be confusing, its output_of_the_model & not model itself. As you said, this model is loaded only once.